lib/Conversion/HandshakeToHW/HandshakeToHW.cpp

```diff
@@ -35,6 +37,45 @@ using namespace circt::hw;
 
 using NameUniquer = std::function<std::string(Operation *)>;
 
+// Shared state used by various functions; captured in a struct to reduce the
+// number of arguments that we have to pass around.
+struct HandshakeLoweringState {
+  ModuleOp parentModule;
+  NameUniquer nameUniquer;
+};
+
+// A type converter is needed to perform the in-flight materialization of "raw"
+// (non-ESI channel) types to their ESI channel correspondents. This comes into
+// effect when backedges exist in the input IR.
+class ESITypeConverter : public TypeConverter {
+public:
+  ESITypeConverter() {
+    addConversion([](Type type) -> Type {
+      if (isa<esi::ChannelType>(type))
+        return type;
+      return esi::ChannelType::get(type.getContext(), type);
```

Comment:
> It might make sense to pass the esiWrapper function in here. Somehow the static functions treats the index type differently.

Which should obviously not be the case. The implementation should be shared across all type conversions - fixed.